### ISSUE TITLE
[QUESTION] Applying Deep Convolutional GAN to other datasets

### ISSUE THREAD

**Issue by Reisa14** [ageron/handson-ml2#552]
**Describe what is unclear to you**
I am trying to apply the Deep Convolutional GAN example from Chapter 17 p. 599 to the fruits-360 dataset on Kaggle:
**Dataset properties**
The total number of images: 90483.
Training set size: 67692 images (one fruit or vegetable per image).
Test set size: 22688 images (one fruit or vegetable per image).
Image size: 100x100 pixels.

I am getting outputs from the following code, but they don't look like fruit at all. My thought process was that in the generator in the example we set the first Dense layer to 7 * 7 * 128 because we have downsampled our 28 x 28 images to 7 x 7 with 128 channels, so I went with 25 x 25 x 128 as the Fruits images are 100 x 100 pixels. My first question is why do we use 128 channels instead of just 1 for greyscale?

We then go down to 64 channels for the Conv2DTranspose to reduce the depth and the stride of 2 upsamples to 14 x 14 (50 x 50 for the fruits images), before having the depth reduced to 1 and being upsampled again to 28 x 28 (100 x 100 for fruits). My second question is why did we settle on a kernel_size of 5?

My next question is why, in the following code, do we multiply by 2 and subtract 1?
```python
X_train_dcgan = x_train.reshape(-1, 100, 100, 1) * 2. - 1.
```

And how do we pick the codings_size and batch_size?

**To Reproduce**
```python
tf.random.set_seed(42)
np.random.seed(42)

codings_size = 100

generator = keras.models.Sequential([
    keras.layers.Dense(25 * 25 * 128, input_shape=[codings_size]),
    keras.layers.Reshape([25, 25, 128]),
    keras.layers.BatchNormalization(),
    keras.layers.Conv2DTranspose(64, kernel_size=5, strides=2, padding="SAME", activation="selu"),
    keras.layers.BatchNormalization(),
    keras.layers.Conv2DTranspose(1, kernel_size=5, strides=2, padding="SAME", activation="tanh"),
])
discriminator = keras.models.Sequential([
    keras.layers.Conv2D(64, kernel_size=5, strides=2, padding="SAME",
                       activation=keras.layers.LeakyReLU(0.2), input_shape=[100, 100, 1]),
    keras.layers.Dropout(0.4),
    keras.layers.Conv2D(128, kernel_size=5, strides=2, padding="SAME",
                       activation=keras.layers.LeakyReLU(0.2)),
    keras.layers.Dropout(0.4),
    keras.layers.Flatten(),
    keras.layers.Dense(1, activation="sigmoid")
])
gan = keras.models.Sequential([generator, discriminator])
```
```python
X_train_dcgan = x_train.reshape(-1, 100, 100, 1) * 2. - 1. # reshape and rescale
```
```python
batch_size = 32
dataset = tf.data.Dataset.from_tensor_slices(X_train_dcgan)
dataset = dataset.shuffle(1000)
dataset = dataset.batch(batch_size, drop_remainder=True).prefetch(1)
```
```python
train_gan(gan, dataset, batch_size, codings_size)
```
Sorry for all the questions, I am trying to make sure I understand what all of the hyperparameters do and whether or not I have made mistakes in my code or just need to tune it!

Thanks

**Comment by ageron** [comment]
Hi @Reisa14,

Thanks for your feedback.

> My first question is why do we use 128 channels instead of just 1 for greyscale?

The generator's input is not an image, it's a dense vector (codings), and the generator's role is to turn it into an image. It's similar to the role of a *decoder* in an Autoencoder: it must decode some latent representation into an image. The codings have fairly few dimensions (e.g., 100), so they can't store information about every detail in the image, just the general aspect. Since the generator will need to build an image, we'll want to use Conv2D layers. But these take images as inputs, and all we have is a vector. So the first thing we need to do is to project this vector into 3D space (height, width, channels). But it's probably best to start by projecting them to a fairly low-resolution image (e.g., 7 x 7), since the codings can't hold much info. The generator's role will be to add the necessary details to make the image realistic. But we have more channels if we want: this way, all of the coding's information will be mapped to "semantic" information for each part of the image. For example, it may encode the fact that the top-left part of the image contains a leaf. It's fairly high level information. Then the Conv2DTranspose layers can increase the spatial resolution and gradually convert the high-level info into low-level information (i.e., pixel colors). If the images are monochrome, you want the last layer to have just 1 filter. If it's RGB, it should be 3.

> My second question is why did we settle on a kernel_size of 5?

I usually use kernel size of 3, except sometimes in the very first layer of a CNN. I may have just tried out various combinations and found that this worked for me. I honestly can't remember. You can just try different kernel sizes and see what works best.

> My next question is why, in the following code, do we multiply by 2 and subtract 1?
The inputs had values ranging from 0 to 1, and I wanted them to range from -1 to +1. Multiplying by 2 and subtracting 1 will do that. This is because neural nets tend to prefer when their inputs have a mean close to 0.

Hope this helps.

**Comment by Reisa14** [comment]
Thanks @ageron , that helps a lot!
